CI is failing:
```
[ERROR] io.quarkus.logging.JsonFormatterTest  Time elapsed: 0.009 s  <<< ERROR!
java.lang.ExceptionInInitializerError
	at sun.misc.Unsafe.ensureClassInitialized(Native Method)
	at sun.reflect.UnsafeFieldAccessorFactory.newFieldAccessor(UnsafeFieldAccessorFactory.java:43)
	at sun.reflect.ReflectionFactory.newFieldAccessor(ReflectionFactory.java:156)
	at java.lang.reflect.Field.acquireFieldAccessor(Field.java:1088)
	at java.lang.reflect.Field.getFieldAccessor(Field.java:1069)
	at java.lang.reflect.Field.get(Field.java:393)
	at org.junit.platform.commons.util.ReflectionUtils.lambda$tryToReadFieldValue$5(ReflectionUtils.java:621)
	at org.junit.platform.commons.function.Try.lambda$call$0(Try.java:57)
	at org.junit.platform.commons.function.Try.of(Try.java:93)
	at org.junit.platform.commons.function.Try.call(Try.java:57)
	at org.junit.platform.commons.util.ReflectionUtils.tryToReadFieldValue(ReflectionUtils.java:621)
	at org.junit.jupiter.engine.descriptor.ExtensionUtils.lambda$registerExtensionsFromFields$3(ExtensionUtils.java:109)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.jupiter.engine.descriptor.ExtensionUtils.registerExtensionsFromFields(ExtensionUtils.java:104)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.prepare(ClassBasedTestDescriptor.java:147)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.prepare(ClassBasedTestDescriptor.java:77)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$prepare$1(NodeTestTask.java:107)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.prepare(NodeTestTask.java:107)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:75)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:188)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:202)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:181)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:142)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:117)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: java.lang.NullPointerException
	at java.util.Properties$LineReader.readLine(Properties.java:434)
	at java.util.Properties.load0(Properties.java:353)
	at java.util.Properties.load(Properties.java:341)
	at io.quarkus.test.QuarkusUnitTest.withConfigurationResource(QuarkusUnitTest.java:398)
	at io.quarkus.logging.JsonFormatterTest.<clinit>(JsonFormatterTest.java:25)
	... 44 more
```

--- a/core/test-extension/deployment/src/main/java/io/quarkus/extest/deployment/TestProcessor.java
+++ b/core/test-extension/deployment/src/main/java/io/quarkus/extest/deployment/TestProcessor.java
@@ -45,6 +45,7 @@ import io.quarkus.deployment.annotations.Record;
 import io.quarkus.deployment.builditem.CapabilityBuildItem;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
 import io.quarkus.deployment.builditem.LaunchModeBuildItem;
+import io.quarkus.deployment.builditem.LogHandlerBuildItem;
 import io.quarkus.deployment.builditem.ObjectSubstitutionBuildItem;
 import io.quarkus.deployment.builditem.ServiceStartBuildItem;
 import io.quarkus.deployment.builditem.ShutdownContextBuildItem;
@@ -64,6 +65,7 @@ import io.quarkus.extest.runtime.config.TestBuildTimeConfig;
 import io.quarkus.extest.runtime.config.TestConfigRoot;
 import io.quarkus.extest.runtime.config.TestRunTimeConfig;
 import io.quarkus.extest.runtime.config.XmlConfig;
+import io.quarkus.extest.runtime.logging.AdditionalLogHandlerValueFactory;
 import io.quarkus.extest.runtime.subst.DSAPublicKeyObjectSubstitution;
 import io.quarkus.extest.runtime.subst.KeyProxy;
 import io.quarkus.runtime.RuntimeValue;
@@ -107,6 +109,17 @@ public final class TestProcessor {
     @BuildStep
     BeanDefiningAnnotationBuildItem registerBeanDefiningAnnotations() {
         return new BeanDefiningAnnotationBuildItem(TEST_ANNOTATION, TEST_ANNOTATION_SCOPE);
+    }
+
+    /**
+     * Register an additional log handler
+     *
+     * @return LogHandlerBuildItem
+     */
+    @BuildStep
+    @Record(RUNTIME_INIT)
+    LogHandlerBuildItem registerAdditionalLogHandler(final AdditionalLogHandlerValueFactory factory) {
+        return new LogHandlerBuildItem(factory.create());
     }
 
     @BuildStep

--- a/core/test-extension/deployment/src/main/resources/application-additional-handlers.properties
+++ b/core/test-extension/deployment/src/main/resources/application-additional-handlers.properties
@@ -1,0 +1,6 @@
+quarkus.log.level=INFO
+quarkus.log.console.enable=true
+quarkus.log.console.level=WARNING
+quarkus.log.console.format=%d{yyyy-MM-dd HH:mm:ss,SSS} %-5p [%c{3.}] (%t) %s%e%n
+# Resource path to DSAPublicKey base64 encoded bytes
+quarkus.root.dsa-key-location=/DSAPublicKey.encoded

--- a/core/test-extension/deployment/src/test/java/io/quarkus/logging/AdditionalHandlersTest.java
+++ b/core/test-extension/deployment/src/test/java/io/quarkus/logging/AdditionalHandlersTest.java
@@ -1,0 +1,43 @@
+package io.quarkus.logging;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.Arrays;
+import java.util.logging.Handler;
+import java.util.logging.Level;
+import java.util.logging.LogManager;
+import java.util.logging.Logger;
+
+import org.jboss.logmanager.handlers.DelayedHandler;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.extest.runtime.logging.AdditionalLogHandlerValueFactory.TestHandler;
+import io.quarkus.runtime.logging.InitialConfigurator;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class AdditionalHandlersTest {
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withConfigurationResource("application-additional-handlers.properties");
+
+    @Test
+    public void additionalHandlersConfigurationTest() {
+        LogManager logManager = LogManager.getLogManager();
+        assertThat(logManager).isInstanceOf(org.jboss.logmanager.LogManager.class);
+
+        DelayedHandler delayedHandler = InitialConfigurator.DELAYED_HANDLER;
+        assertThat(Logger.getLogger("").getHandlers()).contains(delayedHandler);
+        assertThat(delayedHandler.getLevel()).isEqualTo(Level.ALL);
+
+        Handler handler = Arrays.stream(delayedHandler.getHandlers())
+                .filter(h -> (h instanceof TestHandler))
+                .findFirst().orElse(null);
+        assertThat(handler).isNotNull();
+        assertThat(handler.getLevel()).isEqualTo(Level.FINE);
+
+        TestHandler testHandler = (TestHandler) handler;
+        assertThat(testHandler.records).isNotEmpty();
+    }
+
+}

--- a/core/test-extension/runtime/src/main/java/io/quarkus/extest/runtime/logging/AdditionalLogHandlerValueFactory.java
+++ b/core/test-extension/runtime/src/main/java/io/quarkus/extest/runtime/logging/AdditionalLogHandlerValueFactory.java
@@ -1,0 +1,43 @@
+package io.quarkus.extest.runtime.logging;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.logging.Handler;
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+
+import io.quarkus.runtime.RuntimeValue;
+import io.quarkus.runtime.annotations.Recorder;
+
+@Recorder
+public class AdditionalLogHandlerValueFactory {
+
+    public RuntimeValue<Optional<Handler>> create() {
+        return new RuntimeValue<>(Optional.of(new TestHandler()));
+    }
+
+    public static class TestHandler extends Handler {
+
+        public final List<LogRecord> records = new ArrayList<>();
+
+        @Override
+        public void publish(LogRecord record) {
+            records.add(record);
+        }
+
+        @Override
+        public void flush() {
+        }
+
+        @Override
+        public Level getLevel() {
+            return Level.FINE;
+        }
+
+        @Override
+        public void close() throws SecurityException {
+
+        }
+    }
+}

--- a/docs/src/main/asciidoc/logging.adoc
+++ b/docs/src/main/asciidoc/logging.adoc
@@ -40,7 +40,7 @@ The root logger category is handled separately, and is configured via the follow
 |quarkus.log.level|INFO|The default minimum log level for every log category.
 |===
 
-[id="format_string"]
+[id="format-string"]
 == Format String
 
 The logging format string supports the following symbols:
@@ -75,14 +75,14 @@ The logging format string supports the following symbols:
 |%x|Nested Diagnostics context values|Renders all the values from Nested Diagnostics Context in format {value1.value2}
 |===
 
-[id="alt_console_format"]
+[id="alt-console-format"]
 === Alternative Console Logging Formats
 
 It is possible to change the output format of the console log. This can be useful in environments where the output
 of the Quarkus application is captured by a service which can, for example, process and store the log information for
 later analysis.
 
-[id="json_logging"]
+[id="json-logging"]
 ==== JSON Logging
 
 In order to configure JSON logging, the `quarkus-logging-json` extension may be employed.  Add this extension to your

--- a/docs/src/main/asciidoc/logging.adoc
+++ b/docs/src/main/asciidoc/logging.adoc
@@ -75,6 +75,57 @@ The logging format string supports the following symbols:
 |%x|Nested Diagnostics context values|Renders all the values from Nested Diagnostics Context in format {value1.value2}
 |===
 
+[id="alt_console_format"]
+=== Alternative Console Logging Formats
+
+It is possible to change the output format of the console log. This can be useful in environments where the output
+of the Quarkus application is captured by a service which can, for example, process and store the log information for
+later analysis.
+
+[id="json_logging"]
+==== JSON Logging
+
+In order to configure JSON logging, the `quarkus-logging-json` extension may be employed.  Add this extension to your
+application POM as the following snippet illustrates.
+
+.Modifications to POM file to add the JSON logging extension
+[source,xml]
+----
+<build>
+  <dependencies>
+    <!-- ... your other dependencies are here ... -->
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-logging-json</artifactId>
+    </dependency>
+  </dependencies>
+</build>
+----
+
+The presence of this extension will, by default, replace the output format configuration from the console configuration.
+This means that the format string and the color settings (if any) will be ignored.  The other console configuration items
+(including those controlling asynchronous logging and the log level) will continue to be applied.
+
+===== Configuration
+
+The JSON logging extension can be configured in various ways.  The following properties are supported:
+
+[cols="<m,<m,<2",options="header"]
+|===
+|Property Name|Default|Description
+|quarkus.log.json|true|Determine whether JSON logging is enabled; disabling this property will allow the normal console logging to function.
+|quarkus.log.json.pretty-print|false|Determine whether the JSON output is "pretty-printed".
+|quarkus.log.json.date-format|default|Specify the date format to use; `default` indicates that the default format is used.
+|quarkus.log.json.record-delimiter||Specify a record delimiter to add. By default, no delimiter is used.
+|quarkus.log.json.zone-id|default|Specify the time zone ID to use for logging dates; `default` indicates that the default is used.
+|quarkus.log.json.exception-output-type|detailed|Specify the exception output type; values of `detailed`, `formatted`, or `detailed-and-formatted` are allowed.
+|quarkus.log.json.print-details|false|Specify whether detailed caller information should be logged.
+|===
+
+WARNING: Enabling pretty printing might cause certain processors and JSON parsers to fail.
+
+NOTE: Printing the details can be expensive as the values are retrieved from the caller. The details include the
+source class name, source file name, source method name and source line number.
 
 == Examples
 

--- a/extensions/logging-json/deployment/pom.xml
+++ b/extensions/logging-json/deployment/pom.xml
@@ -23,6 +23,27 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-logging-json</artifactId>
         </dependency>
+        <!-- test dependencies -->
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-junit5-internal</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-junit5</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-arc-deployment</artifactId>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
     <build>

--- a/extensions/logging-json/deployment/src/test/java/io/quarkus/logging/json/JsonFormatterCustomConfigTest.java
+++ b/extensions/logging-json/deployment/src/test/java/io/quarkus/logging/json/JsonFormatterCustomConfigTest.java
@@ -1,0 +1,56 @@
+package io.quarkus.logging.json;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.time.ZoneId;
+import java.util.Arrays;
+import java.util.logging.Formatter;
+import java.util.logging.Handler;
+import java.util.logging.Level;
+import java.util.logging.LogManager;
+import java.util.logging.Logger;
+
+import org.jboss.logmanager.formatters.JsonFormatter;
+import org.jboss.logmanager.formatters.StructuredFormatter;
+import org.jboss.logmanager.handlers.ConsoleHandler;
+import org.jboss.logmanager.handlers.DelayedHandler;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.runtime.logging.InitialConfigurator;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class JsonFormatterCustomConfigTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withConfigurationResource("application-json-formatter-custom.properties");
+
+    @Test
+    public void jsonFormatterCustomConfigurationTest() {
+        LogManager logManager = LogManager.getLogManager();
+        assertThat(logManager).isInstanceOf(org.jboss.logmanager.LogManager.class);
+
+        DelayedHandler delayedHandler = InitialConfigurator.DELAYED_HANDLER;
+        assertThat(Logger.getLogger("").getHandlers()).contains(delayedHandler);
+        assertThat(delayedHandler.getLevel()).isEqualTo(Level.ALL);
+
+        Handler handler = Arrays.stream(delayedHandler.getHandlers())
+                .filter(h -> (h instanceof ConsoleHandler))
+                .findFirst().orElse(null);
+        assertThat(handler).isNotNull();
+        assertThat(handler.getLevel()).isEqualTo(Level.WARNING);
+
+        Formatter formatter = handler.getFormatter();
+        assertThat(formatter).isInstanceOf(JsonFormatter.class);
+        JsonFormatter jsonFormatter = (JsonFormatter) formatter;
+        assertThat(jsonFormatter.isPrettyPrint()).isTrue();
+        assertThat(jsonFormatter.getDateTimeFormatter().toString())
+                .isEqualTo("Value(DayOfMonth)' 'Text(MonthOfYear,SHORT)' 'Value(Year,4,19,EXCEEDS_PAD)");
+        assertThat(jsonFormatter.getDateTimeFormatter().getZone()).isEqualTo(ZoneId.of("UTC+05:00"));
+        assertThat(jsonFormatter.getExceptionOutputType())
+                .isEqualTo(StructuredFormatter.ExceptionOutputType.DETAILED_AND_FORMATTED);
+        assertThat(jsonFormatter.getRecordDelimiter()).isEqualTo("\n;");
+        assertThat(jsonFormatter.isPrintDetails()).isTrue();
+    }
+}

--- a/extensions/logging-json/deployment/src/test/java/io/quarkus/logging/json/JsonFormatterDefaultConfigTest.java
+++ b/extensions/logging-json/deployment/src/test/java/io/quarkus/logging/json/JsonFormatterDefaultConfigTest.java
@@ -1,0 +1,56 @@
+package io.quarkus.logging.json;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Arrays;
+import java.util.logging.Formatter;
+import java.util.logging.Handler;
+import java.util.logging.Level;
+import java.util.logging.LogManager;
+import java.util.logging.Logger;
+
+import org.jboss.logmanager.formatters.JsonFormatter;
+import org.jboss.logmanager.formatters.StructuredFormatter;
+import org.jboss.logmanager.handlers.ConsoleHandler;
+import org.jboss.logmanager.handlers.DelayedHandler;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.runtime.logging.InitialConfigurator;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class JsonFormatterDefaultConfigTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withConfigurationResource("application-json-formatter-default.properties");
+
+    @Test
+    public void jsonFormatterDefaultConfigurationTest() {
+        LogManager logManager = LogManager.getLogManager();
+        assertThat(logManager).isInstanceOf(org.jboss.logmanager.LogManager.class);
+
+        DelayedHandler delayedHandler = InitialConfigurator.DELAYED_HANDLER;
+        assertThat(Logger.getLogger("").getHandlers()).contains(delayedHandler);
+        assertThat(delayedHandler.getLevel()).isEqualTo(Level.ALL);
+
+        Handler handler = Arrays.stream(delayedHandler.getHandlers())
+                .filter(h -> (h instanceof ConsoleHandler))
+                .findFirst().orElse(null);
+        assertThat(handler).isNotNull();
+        assertThat(handler.getLevel()).isEqualTo(Level.WARNING);
+
+        Formatter formatter = handler.getFormatter();
+        assertThat(formatter).isInstanceOf(JsonFormatter.class);
+        JsonFormatter jsonFormatter = (JsonFormatter) formatter;
+        assertThat(jsonFormatter.isPrettyPrint()).isFalse();
+        assertThat(jsonFormatter.getDateTimeFormatter().toString())
+                .isEqualTo(DateTimeFormatter.ISO_OFFSET_DATE_TIME.withZone(ZoneId.systemDefault()).toString());
+        assertThat(jsonFormatter.getDateTimeFormatter().getZone()).isEqualTo(ZoneId.systemDefault());
+        assertThat(jsonFormatter.getExceptionOutputType()).isEqualTo(StructuredFormatter.ExceptionOutputType.DETAILED);
+        assertThat(jsonFormatter.getRecordDelimiter()).isEqualTo("\n");
+        assertThat(jsonFormatter.isPrintDetails()).isFalse();
+    }
+}

--- a/extensions/logging-json/deployment/src/test/resources/application-json-formatter-custom.properties
+++ b/extensions/logging-json/deployment/src/test/resources/application-json-formatter-custom.properties
@@ -1,0 +1,11 @@
+quarkus.log.level=INFO
+quarkus.log.console.enable=true
+quarkus.log.console.level=WARNING
+quarkus.log.console.json=true
+quarkus.log.console.json.pretty-print=true
+quarkus.log.console.json.date-format=d MMM uuuu
+quarkus.log.console.json.record-delimiter=\n;
+quarkus.log.console.json.zone-id=UTC+05:00
+quarkus.log.console.json.exception-output-type=DETAILED_AND_FORMATTED
+quarkus.log.console.json.print-details=true
+

--- a/extensions/logging-json/deployment/src/test/resources/application-json-formatter-default.properties
+++ b/extensions/logging-json/deployment/src/test/resources/application-json-formatter-default.properties
@@ -1,0 +1,5 @@
+quarkus.log.level=INFO
+quarkus.log.console.enable=true
+quarkus.log.console.level=WARNING
+quarkus.log.console.format=%d{yyyy-MM-dd HH:mm:ss,SSS} %-5p [%c{3.}] (%t) %s%e%n
+quarkus.log.console.json=true

